examples/create_events_with_tpl_mutex.rs

```diff
@@ -0,0 +1,101 @@
+extern crate alloc;
+
+use alloc::boxed::Box;
+use core::{ffi::c_void, mem::MaybeUninit, ptr};
+
+use mu_rust_helpers::{
+  boot_services::{event::EventType, tpl::Tpl, BootServices, StandardBootServices},
+  tpl_mutex::TplMutex,
+};
+
+use r_efi::efi;
+
+#[derive(Debug)]
+struct MyContext {
+  _some_immutable_state: usize,
+  _some_other_immutable_state: efi::Handle,
+  some_mutable_state: TplMutex<'static, i32>,
+  _some_other_mutable_state: TplMutex<'static, String>,
+}
+unsafe impl Sync for MyContext {}
+
+extern "efiapi" fn event_notify_callback_tpl_mutex(_event: efi::Event, context: &'static MyContext) {
+  let mut some_mutable_state = context.some_mutable_state.lock();
+  *some_mutable_state += 1;
+}
+
+extern "efiapi" fn event_notify_callback_tpl_mutex_2(_event: efi::Event, context: Option<&'static MyContext>) {
+  println!("{context:?}")
+}
+
+extern "efiapi" fn event_notify_callback_void(_event: efi::Event, context: Box<()>) {
+  println!("{context:?}")
+}
+
+fn main() {
+  static BOOT_SERVICE: StandardBootServices = StandardBootServices::new_uninit();
+  let efi_boot_services = unsafe {
+    let mut bs = MaybeUninit::<efi::BootServices>::zeroed();
+    bs.assume_init_mut().create_event = efi_create_event;
+    bs.assume_init_mut().raise_tpl = efi_raise_tpl;
+    bs.assume_init_mut().restore_tpl = efi_restore_tpl;
+    bs.assume_init()
+  };
+  BOOT_SERVICE.initialize(&efi_boot_services);
+
+  let ctx = Box::new(MyContext {
+    _some_immutable_state: 0,
+    _some_other_immutable_state: ptr::null_mut(),
+    some_mutable_state: TplMutex::new(&BOOT_SERVICE, Tpl::APPLICATION, 0),

```

Comment:
nitpick: it doesn't make sense to "raise" TPL to Tpl::APPLICATION, since it is the lowest TPL. I think it is _technically_ legal, but it would not ever make sense in practice. recommend using CALLBACK or NOTIFY in the example just so it is more typical usage.